### DISCUSSION TITLE
TMP_HLQ - Lineinfile

### DISCUSSION THREAD

**Discussion by ghost** [ansible-collections/ibm_zos_core#834] [Q&A]
Hi guys!
I'm using the new parameter tmp_hlq in zos_lineinfile but I'm still having problems executing an automation. I couldn't confirm if the automation is using this hlq, since the error shows the file name without the selected hlq.

![image](https://github.com/ansible-collections/ibm_zos_core/assets/105386204/55098924-dabb-468b-8600-b34d2939fc00)

Another thing I would like to confirm is: Can I use more than one qualifier in HLQ? As? In my environment I would like to use: "DSS.CIC", but it does not accept it and gives an error in the qualifier:

![image](https://github.com/ansible-collections/ibm_zos_core/assets/105386204/2a6a98c0-a59f-4bc5-87f9-7178eed36782)

Note: In the lab environment, I was able to solve the problem by deleting the tmp_hlq parameter and granting access to the master catalog to the automation user, but in the production environment this access is blocked.

My environment is using ZOAU 1.2 and zos_core 1.6.0

**Comment by ghost** [comment]
Just to complement the information, I found an error in the zOS log, indicating that the automation is trying to access the master catalog, but if it was using the qualifier inserted in tmp_hlq and it was trying to write DST.ASBCIC01.** instead of ASBCIC01.* * this would not be happening.

![image](https://github.com/ansible-collections/ibm_zos_core/assets/105386204/7b91eeef-fef5-44e4-b4e4-c0e9dfd0985e)

**Reply by ddimatos** [reply]
Hi @fabiobncbb , I'll try to recreate but I'll do so on a supported version of zoau. Ansible z/OS core 1.6.0 requires minimal zoau 1.2.2.

We don't perform recreates on unsupported combinations.

**Comment by ddimatos** [comment]
Hi, I will answer the second question first, currently our tmp_hlq does not allow for more than one single segment meaning it must be no more than 8 characters long allowing for [A-Z]{1}[A-Z0-9]{0,7}. 
1) I see no reason we could not allow for more then one segment so long as the the tmp_hlq + remaining were validated before processing.
2) The tmp_hlq has left out some special chars so you are limited to `[A-Z]{1}[A-Z0-9]{0,7}`.

If this is something you wish to be supported, please open an "Enhancement" issue so this can be triaged and then prioritized against other current and backlog entries.

**Comment by ddimatos** [comment]
I have opened issue https://github.com/ansible-collections/ibm_zos_core/issues/909 to correct some of my findings, please review it when you can.

**Comment by ddimatos** [comment]
To address your first issue, although I did not see your playbook, my guess is that you are using `backup_name` and your source is a data set.

For now, your work around is:
1) Use zos_copy to perform a backup and don't use the one built into zos_blockinfile
2) Use zos_backup_restore module
3) Don't use the `backup_name`, set only `backup:true` and set tmp_hlq, this will allow the module to use tmp_hlq and for the remaining qualifiers produce random values. This only works if your source is a Sequential data set, not a PDS/E.
```
    - name: Update Data set using Marker
      zos_blockinfile:
        src: "ANSIBLE.CKCARLA.OUT1"
        marker: "/* {mark} ANSIBLE MANAGED BLOCK */"
        insertafter: "EOF"
        block: |
          MOUNT FILESYSTEM('FOO.DATA.SET') TYPE(ZFS) MODE(READ)
            MOUNTPOINT('/tmp/src/somedirectory')
        tmp_hlq: "SUPER"
        backup: true
```
Results with backup_name
```
"backup_name": "SUPER.P3554478.T0878090.C0000000",
```
4) If you would like to try out an unsupported hotfix, it will allow for you to have a src that is a sequential dataset where you can use backup_name:
Playbook:
```
    - name: Update Data set with Marker
      zos_blockinfile:
        src: "ANSIBLE.CKCARLA.OUT1"
        marker: "/* {mark} ANSIBLE MANAGED BLOCK */"
        insertafter: "EOF"
        block: |
          MOUNT FILESYSTEM('FOO.DATA.SET') TYPE(ZFS) MODE(READ)
            MOUNTPOINT('/tmp/src/somedirectory')
        tmp_hlq: "SUPER"
        backup: true
        backup_name: "ANSIBLE.BCKUP.LLQ1"
      register: result
```

Output:
```
"backup_name": "SUPER.P3554557.T0162263.C0000000",
```
see issue #910
[ibm-ibm_zos_core-1.6.0.tar.gz](https://github.com/ansible-collections/ibm_zos_core/files/12138250/ibm-ibm_zos_core-1.6.0.tar.gz)

**Comment by ghost** [comment]
Hi!

I'm not using backup and I'm getting the same error accessing the master catalog, both in the zos_lineinfile and zos_copy modules, using or not using the tmp_hlq parameter.
Below is an example of my task and the error returned:

This is my task

  zos_copy:
    src: "{{ hlq_amb }}.PLEX{{ hlq_lib }}.PROCLIB.CTG({{ cics_modelo }})"
    dest: "{{ hlq_amb }}.PLEX{{ hlq_lib }}.PROCLIB.CTG({{ cics_novo }})"
    remote_src: true
    force: true
    backup: true  ## (I tried with and without this parameter)
    tmp_hlq: DST  ## (I tried with and without this parameter)

This is the result with -vvvv

TASK [cria_proclib : Faz uma cópia da proc do cics modelo e salva com o nome do cics novo] *******************************************************************************************************************************
task path: /git/ansible/roles/cria_proclib/tasks/main.yml:2
<desb-des-asb.bb.com.br> ESTABLISH SSH CONNECTION FOR USER: asbcic01
<desb-des-asb.bb.com.br> SSH: EXEC sshpass -d10 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="asbcic01"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/f3160882/.ansible/cp/bf322e5ee0"' desb-des-asb.bb.com.br '/bin/sh -c '"'"'echo ~asbcic01 && sleep 0'"'"''
<desb-des-asb.bb.com.br> (0, b'/home/asbcic01\n', b'OpenSSH_8.0p1, OpenSSL 1.1.1k  FIPS 25 Mar 2021\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 3: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 5 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 2427484\r\ndebug3: mux_client_request_session: session request sent\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<desb-des-asb.bb.com.br> ESTABLISH SSH CONNECTION FOR USER: asbcic01
<desb-des-asb.bb.com.br> SSH: EXEC sshpass -d10 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="asbcic01"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/f3160882/.ansible/cp/bf322e5ee0"' desb-des-asb.bb.com.br '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo /home/asbcic01/.ansible/tmp `"&& mkdir "` echo /home/asbcic01/.ansible/tmp/ansible-tmp-1690214863.102096-2429389-224800459169884 `" && echo ansible-tmp-1690214863.102096-2429389-224800459169884="` echo /home/asbcic01/.ansible/tmp/ansible-tmp-1690214863.102096-2429389-224800459169884 `" ) && sleep 0'"'"''
<desb-des-asb.bb.com.br> (0, b'ansible-tmp-1690214863.102096-2429389-224800459169884=/home/asbcic01/.ansible/tmp/ansible-tmp-1690214863.102096-2429389-224800459169884\n', b'OpenSSH_8.0p1, OpenSSL 1.1.1k  FIPS 25 Mar 2021\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 3: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 5 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 2427484\r\ndebug3: mux_client_request_session: session request sent\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
Using module file /git/ansible/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_copy.py
<desb-des-asb.bb.com.br> PUT /home/f3160882/.ansible/tmp/ansible-local-2429235jouabdsx/tmpqdm_tlc3 TO /home/asbcic01/.ansible/tmp/ansible-tmp-1690214863.102096-2429389-224800459169884/AnsiballZ_zos_copy.py
<desb-des-asb.bb.com.br> SSH: EXEC sshpass -d10 sftp -o BatchMode=no -b - -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="asbcic01"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/f3160882/.ansible/cp/bf322e5ee0"' '[desb-des-asb.bb.com.br]'
<desb-des-asb.bb.com.br> (0, b'sftp> put /home/f3160882/.ansible/tmp/ansible-local-2429235jouabdsx/tmpqdm_tlc3 /home/asbcic01/.ansible/tmp/ansible-tmp-1690214863.102096-2429389-224800459169884/AnsiballZ_zos_copy.py\n', b'OpenSSH_8.0p1, OpenSSL 1.1.1k  FIPS 25 Mar 2021\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 3: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 5 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 2427484\r\ndebug3: mux_client_request_session: session request sent\r\ndebug2: Remote version: 3\r\ndebug2: Server supports extension "posix-rename@openssh.com" revision 1\r\ndebug2: Server supports extension "statvfs@openssh.com" revision 2\r\ndebug2: Server supports extension "fstatvfs@openssh.com" revision 2\r\ndebug2: Server supports extension "hardlink@openssh.com" revision 1\r\ndebug2: Server supports extension "fsync@openssh.com" revision 1\r\ndebug3: Sent message fd 3 T:16 I:1\r\ndebug3: SSH_FXP_REALPATH . -> /DESB/home/asbcic01 size 0\r\ndebug3: Looking up /home/f3160882/.ansible/tmp/ansible-local-2429235jouabdsx/tmpqdm_tlc3\r\ndebug3: Sent message fd 3 T:17 I:2\r\ndebug3: Received stat reply T:101 I:2\r\ndebug1: Couldn\'t stat remote file: No such file or directory\r\ndebug3: Sent message SSH2_FXP_OPEN I:3 P:/home/asbcic01/.ansible/tmp/ansible-tmp-1690214863.102096-2429389-224800459169884/AnsiballZ_zos_copy.py\r\ndebug3: Sent message SSH2_FXP_WRITE I:4 O:0 S:32768\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 4 32768 bytes at 0\r\ndebug3: Sent message SSH2_FXP_WRITE I:5 O:32768 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:6 O:65536 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:7 O:98304 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:8 O:131072 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:9 O:163840 S:32768\r\ndebug3: Sent message SSH2_FXP_WRITE I:10 O:196608 S:17829\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 5 32768 bytes at 32768\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 6 32768 bytes at 65536\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 7 32768 bytes at 98304\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 8 32768 bytes at 131072\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 9 32768 bytes at 163840\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: In write loop, ack for 10 17829 bytes at 196608\r\ndebug3: Sent message SSH2_FXP_CLOSE I:4\r\ndebug3: SSH2_FXP_STATUS 0\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<desb-des-asb.bb.com.br> ESTABLISH SSH CONNECTION FOR USER: asbcic01
<desb-des-asb.bb.com.br> SSH: EXEC sshpass -d10 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="asbcic01"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/f3160882/.ansible/cp/bf322e5ee0"' desb-des-asb.bb.com.br '/bin/sh -c '"'"'chmod u+x /home/asbcic01/.ansible/tmp/ansible-tmp-1690214863.102096-2429389-224800459169884/ /home/asbcic01/.ansible/tmp/ansible-tmp-1690214863.102096-2429389-224800459169884/AnsiballZ_zos_copy.py && sleep 0'"'"''
<desb-des-asb.bb.com.br> (0, b'', b'OpenSSH_8.0p1, OpenSSL 1.1.1k  FIPS 25 Mar 2021\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 3: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 5 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 2427484\r\ndebug3: mux_client_request_session: session request sent\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
<desb-des-asb.bb.com.br> ESTABLISH SSH CONNECTION FOR USER: asbcic01
<desb-des-asb.bb.com.br> SSH: EXEC sshpass -d10 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="asbcic01"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/f3160882/.ansible/cp/bf322e5ee0"' -tt desb-des-asb.bb.com.br '/bin/sh -c '"'"'_BPXK_AUTOCVT=ON ZOAU_HOME=/pp/zoau/v1r2m0 PYTHONPATH=/pp/zoau/v1r2m0/lib LIBPATH=/pp/zoau/v1r2m0/lib:/pp/python/v3r9m0/pyz/bin/python3.9/lib:/lib:/usr/lib:. PATH=/pp/zoau/v1r2m0/bin:/pp/python/v3r9m0/pyz/bin/python3.9/bin:/bin:/var/bin _CEE_RUNOPTS='"'"'"'"'"'"'"'"'FILETAG(AUTOCVT,AUTOTAG) POSIX(ON)'"'"'"'"'"'"'"'"' _TAG_REDIR_ERR=txt _TAG_REDIR_IN=txt _TAG_REDIR_OUT=txt LANG=C PYTHONUNBUFFERED=1 /pp/python/v3r9m0/pyz/bin/python3.9 /home/asbcic01/.ansible/tmp/ansible-tmp-1690214863.102096-2429389-224800459169884/AnsiballZ_zos_copy.py && sleep 0'"'"''
<desb-des-asb.bb.com.br> (1, b'\r\n{"failed": true, "msg": "Source data set DSS.PLEXDES.PROCLIB.CTG(CIC0A0Z) does not exist", "exception": "  File \\"/tmp/ansible_ibm.ibm_zos_core.zos_copy_payload_1hqoy2qd/ansible_ibm.ibm_zos_core.zos_copy_payload.zip/ansible_collections/ibm/ibm_zos_core/plugins/modules/zos_copy.py\\", line 2272, in run_module\\n", "invocation": {"module_args": {"src": "DSS.PLEXDES.PROCLIB.CTG(CIC0A0Z)", "dest": "DSS.PLEXDES.PROCLIB.CTG(CICANSI)", "remote_src": true, "force": true, "backup": true, "tmp_hlq": "DST", "is_uss": false, "is_pds": false, "is_src_dir": false, "copy_member": true, "src_member": true, "temp_path": null, "is_mvs_dest": false, "local_charset": "UTF-8", "is_binary": false, "local_follow": true, "ignore_sftp_stderr": false, "validate": false, "unsafe_writes": false, "encoding": null, "content": null, "backup_name": null, "volume": null, "dest_data_set": null, "size": null, "mode": null, "owner": null, "group": null, "seuser": null, "serole": null, "selevel": null, "setype": null, "attributes": null}}}\r\n', b'OpenSSH_8.0p1, OpenSSL 1.1.1k  FIPS 25 Mar 2021\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 3: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 5 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 2427484\r\ndebug3: mux_client_request_session: session request sent\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 1\r\nShared connection to desb-des-asb.bb.com.br closed.\r\n')
<desb-des-asb.bb.com.br> Failed to connect to the host via ssh: OpenSSH_8.0p1, OpenSSL 1.1.1k  FIPS 25 Mar 2021
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 3: Applying options for *
debug1: auto-mux: Trying existing master
debug2: fd 5 setting O_NONBLOCK
debug2: mux_client_hello_exchange: master version 4
debug3: mux_client_forwards: request forwardings: 0 local, 0 remote
debug3: mux_client_request_session: entering
debug3: mux_client_request_alive: entering
debug3: mux_client_request_alive: done pid = 2427484
debug3: mux_client_request_session: session request sent
debug3: mux_client_read_packet: read header failed: Broken pipe
debug2: Received exit status from master 1
Shared connection to desb-des-asb.bb.com.br closed.
<desb-des-asb.bb.com.br> ESTABLISH SSH CONNECTION FOR USER: asbcic01
<desb-des-asb.bb.com.br> SSH: EXEC sshpass -d10 ssh -vvv -C -o ControlMaster=auto -o ControlPersist=60s -o 'User="asbcic01"' -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o 'ControlPath="/home/f3160882/.ansible/cp/bf322e5ee0"' desb-des-asb.bb.com.br '/bin/sh -c '"'"'rm -f -r /home/asbcic01/.ansible/tmp/ansible-tmp-1690214863.102096-2429389-224800459169884/ > /dev/null 2>&1 && sleep 0'"'"''
<desb-des-asb.bb.com.br> (0, b'', b'OpenSSH_8.0p1, OpenSSL 1.1.1k  FIPS 25 Mar 2021\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 3: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 5 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 2427484\r\ndebug3: mux_client_request_session: session request sent\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
fatal: [DESB]: FAILED! => {
    "backup_name": null,
    "changed": false,
    "invocation": {
        "module_args": {
            "backup": true,
            "dest": "DSS.PLEXDES.PROCLIB.CTG(CICANSI)",
            "force": true,
            "remote_src": true,
            "src": "DSS.PLEXDES.PROCLIB.CTG(CIC0A0Z)",
            "tmp_hlq": "DST"
        }
    },
    **"msg": "Source data set DSS.PLEXDES.PROCLIB.CTG(CIC0A0Z) does not exist",** but it exists
    "rc": null,
    "stderr": null,
    "stderr_lines": null,
    "stdout": null,
    "stdout_lines": null
}

This is the error in the zos log, which shouldn't happen if you were using tmp_hlq correctly, because it shouldn't access the master_catalog, but a user_catalog.
Unfortunately here we have very strict rules and no user or application has access to the master_catalog, everyone is directed to the user_catalog with HLQ's, aliases, etc.


ICH408I USER(ASBCIC01) GROUP(OMVSGRP ) NAME(ANSIBLE CICS USER 01) 243
  CATALOG.DES4 CL(DATASET ) VOL(RE0400)                              
  INSUFFICIENT ACCESS AUTHORITY                                      
  FROM CATALOG.%%%% (G)                                              
  ACCESS INTENT(UPDATE )  ACCESS ALLOWED(READ   )         
  
  I'm using ZOA 1.2 and zos_core 1.6

**Reply by ddimatos** [reply]
@fabiobncbb Thank you for this information, it makes troubleshooting easier. 

There are a few things here happening, first lets back up and explain something that that is corrected in 1.7.0-beta.1 coming out in about a week. In 1.5.0 - 1.6.0 because the zos_copy module had been redesigned completely, the team added a 'automatic' backup strategy meaning we performed a backup no matter what even if we were not asked to, then on success removed it; this was a data integrity safety check that we probably should have never introduced, thus its being removed in 1.7.0-beta.1.

In 1.6.0 we allowed users to bypass the automatic backup with `force:true` (which you are doing, keep that enabled)

`tmp_hlq` maybe could use more documentation but it is only meant to give users the ability to override a HLQ on a temporary data set that the user will never see or even know was created and cleaned up. We wanted to give users this power over temporary data sets to prevent errors for when systems limited the HLQ. 

I think the error ""Source data set DSS.PLEXDES.PROCLIB.CTG(CIC0A0Z) does not exist"," is because a LISTCAT was performed to see if the data set existed and was cataloged, your user `asbcic01` tried to access the master catalog and was not permitted, change the `src` to include the HLQ, don't expect `tmp_hlq` to be prepended to `src`. 

For src and dest, you should be setting the HLQ, in other words, `tmp_hlq` is not prepended to your `src` or `dest`,try:
```
src: "DST.{{ hlq_amb }}.PLEX{{ hlq_lib }}.PROCLIB.CTG({{ cics_modelo }})"
dest: "DST.{{ hlq_amb }}.PLEX{{ hlq_lib }}.PROCLIB.CTG({{ cics_novo }})"
remote_src: true
force: true
backup: false 
```

There is temporary data sets used when we invoke LISTCAT, LISTDS, and other VTOC commands, but this time I did not see any error for that. For example in your first `zos_block infile` example there was clear evidence of that:
<img width="1656" alt="image" src="https://github.com/ansible-collections/ibm_zos_core/assets/25803172/49a25388-4236-4a2d-9b4e-8eb34be7ca87">.

I don't need a recreate to see that our commands to call LISTCAT, LISTDS, and other VTOC  are not passing tmp_hlq; **could you try my zos_copy suggestion above** and for zos_blockinfile set the **GLOBAL environment var** `  TMPHLQ:DST`, in your enviroment_vars like so:

```
environment_vars:
  _BPXK_AUTOCVT: "ON"
  ZOAU_HOME: "{{ ZOAU }}"
  PYTHONPATH: "{{ ZOAU }}/lib"
  LIBPATH: "{{ ZOAU }}/lib:{{ PYZ }}/lib:/lib:/usr/lib:."
  PATH: "{{ ZOAU }}/bin:{{ PYZ }}/bin:/bin:/var/bin"
  _CEE_RUNOPTS: "FILETAG(AUTOCVT,AUTOTAG) POSIX(ON)"
  _TAG_REDIR_ERR: "txt"
  _TAG_REDIR_IN: "txt"
  _TAG_REDIR_OUT: "txt"
  LANG: "C"
  TMPHLQ: DST
```

`

**Reply by ghost** [reply]
Hi @ddimatos!

I figured tmp_hlq was just for the temporary files that the user would never need to see or know existed, like these and that he wouldn't put that tmp in front of the scr ever.

![image](https://github.com/ansible-collections/ibm_zos_core/assets/105386204/b7edc171-9f98-4e94-b31b-80074d66ef50)

Unfortunately I can't put the tmp_hlq in the scr, because that would change the path where it would look for the old member and create the new one, as the correct path is DSS and not DST, as below:

Translate the path:
{{ hlq_amb }}.PLEX{{ hlq_lib }}.PROCLIB.CTG({{ model_cics }}) is equal DSS.PLEXDES.PROCLIB.CTG({{ model_cics }}).

DSS is the path to the member and DST was the path just to temprary files ASBCIC01.** (DST.ASBCIC01.**) 

I also couldn't put DSS as tmp_hlq because DSS libraries need much more permissions and access rules than DST (temporary) libraries and since the idea is that they were only used for temporary files that would be deleted at the end of the run, DST would be the best choice.

Today I managed to resolve the situation. As?
Without using the tmp_hlq parameter and creating an alias for the temporary file ASBCIC01.** that diverts the creation of this temporary from the master_catalog to a user_catalog. This is my task today:

   zos_copy:
     src: "{{ hlq_amb }}.PLEX{{ hlq_lib }}.PROCLIB.CTG({{ model_cics }})" ## "DSS.PLEXDES.PROCLIB.CTG({{ model_cics }})"
     dest: "{{ hlq_amb }}.PLEX{{ hlq_lib }}.PROCLIB.CTG({{ cics_new }})"
     remote_src: true
     force: true

The issue is that this alias is different from the standard of our environment (it is not allowed to create aliases for users, only for applications) and I don't know if I will be able to create this alias in the production environment, so I wanted to understand why it is not working, but I also believe that, as a new version of the module is about to be released, it is not worth the effort to discover the cause and we can close this discussion.

**Comment by ddimatos** [comment]
It was a bit difficult to follow without additional details; it looked to me like the error for zos_copy was `source data set DSS.PLEXDES.PROCLIB.CTG(CIC0A0Z) does not exist"` which you noted required special privileges to access thus I assumed LISTCAT performed on it could not without privileges, so the code thought it did not exist. How I interpreted the scenario is that I thought maybe you figured setting `tmp_hlq` would be prepended to the `src` and `dest` directing `zos_copy` to a private proclib because of the `tmp_hlq` prefix, it was a bit of a misunderstanding and assuming on my part without added details. 

In the `zos_copy` backtrace I did not see any errors for temporary files like I saw in `zos_lineinfile`, which lead me to believe you were trying to prepend tmp_hlq to the src and destination. The translated paths help me understand. 

Despite the confusion, **there is an issue, the code that drives the discover of data sets is not taking into consideration the tmp_hlq, that is a bug, and I will open an issue for it.** This impacts nearly every module that is working with a data set, 1.7.0-beta.1 does not address this issue, it will persist there. 

I do not know if you saw my comment but there is a way to achieve a temporary high level qualifier which is what we offered in 1.4.0 and still exists, that is to add the environment variable ` TMPHLQ: DST` to your vars, I shared this above, it is worth trying to allow you to avoid creating an Alias which is a creative work around. 

Ill share what that looks like once more, see the last variable `TMPHLQ: DST` <---- try this, it will be global for all temp files but may suffice for now. 
```
environment_vars:
  _BPXK_AUTOCVT: "ON"
  ZOAU_HOME: "{{ ZOAU }}"
  PYTHONPATH: "{{ ZOAU }}/lib"
  LIBPATH: "{{ ZOAU }}/lib:{{ PYZ }}/lib:/lib:/usr/lib:."
  PATH: "{{ ZOAU }}/bin:{{ PYZ }}/bin:/bin:/var/bin"
  _CEE_RUNOPTS: "FILETAG(AUTOCVT,AUTOTAG) POSIX(ON)"
  _TAG_REDIR_ERR: "txt"
  _TAG_REDIR_IN: "txt"
  _TAG_REDIR_OUT: "txt"
  LANG: "C"
  TMPHLQ: DST
```

**Comment by ddimatos** [comment]
I will close the issue as you suggested but leave it unlocked so updates can be provided. It would be good to know if the environment var workaround `TMPHLQ: DST` worked or not, but to summarize, I believe the cause is in the code we use to validate data sets and opened issue #912
